### PR TITLE
Ps 8.0 4960 4962

### PR DESCRIPTION
**Pull request by george-lorch**

### CODE REVIEW

**Comment by george-lorch** [comment]
https://ps.cd.percona.com/view/8.0/job/percona-server-8.0-param/86/

File: storage/tokudb/hatoku_cmp.cc [hunk 1]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+  // upper bound of key comparison descriptor
+  max_row_desc_buff_size = 2 * (form->s->fields * 6) + 10;
+  // upper bound for sec. key part
+  max_row_desc_buff_size += get_max_secondary_key_pack_desc_size();
+  // upper bound for clustering val part
+  max_row_desc_buff_size += get_max_clustering_val_pack_desc_size(form->s);
+  return max_row_desc_buff_size;
+}
+
+uint32_t KEY_AND_COL_INFO::get_max_secondary_key_pack_desc_size() const {
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: Mark inline, and move to header, it seems it should compile to "return a constant plus a bit". Might as well mark as noexcept.

File: storage/tokudb/hatoku_cmp.cc [hunk 2]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+  uint32_t len = 0;
+  for (uint32_t i = 0; i < table_share->fields; i++) {
+    if (is_variable_field(i) && !bitmap_is_set(&key_filters[keynr], i)) {
+      len += num_offset_bytes;
+    }
+  }
+  return len;
+}
+
+uint32_t KEY_AND_COL_INFO::get_max_desc_size(TABLE *form) const {
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: Mark inline, and move to header, it seems it should compile to "return a constant plus a bit". Might as well mark as noexcept.

File: storage/tokudb/hatoku_cmp.cc [hunk 3]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+  uint32_t offset = 0;
+  for (uint32_t i = 0; i < table_share->fields; i++) {
+    if (is_fixed_field(i) && !bitmap_is_set(&key_filters[keynr], i)) {
+      offset += field_lengths[i];
+    }
+  }
+  return offset;
+}
+
+uint32_t KEY_AND_COL_INFO::get_len_of_offsets(TABLE_SHARE *table_share,
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: consider const ref for table_share

File: storage/tokudb/hatoku_cmp.cc [hunk 4]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+    num_offset_bytes = 0;
+
+    allocated = false;
+  }
+}
+
+void KEY_AND_COL_INFO::reset(uint32_t keynr) {
+  bitmap_clear_all(&key_filters[keynr]);
+  tokudb::memory::free(cp_info[keynr]);
+  cp_info[keynr] = NULL;
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: nullptr

File: storage/tokudb/hatoku_cmp.cc [hunk 5]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+
+void KEY_AND_COL_INFO::free() {
+  if (allocated) {
+    for (uint32_t i = 0; i < MAX_KEY + 1; i++) {
+      bitmap_free(&key_filters[i]);
+    }
+
+    for (uint32_t i = 0; i < MAX_KEY + 1; i++) {
+      tokudb::memory::free(cp_info[i]);
+      cp_info[i] = NULL;  // 3144
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: s/NULL/nullptr/g

File: storage/tokudb/hatoku_cmp.cc [hunk 6]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+    }
+
+    tokudb::memory::free(multi_ptr);
+    field_types = NULL;
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: s/NULL/nullptr/g

File: storage/tokudb/hatoku_cmp.cc [hunk 7]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+  return error;
+}
+
+int32_t KEY_AND_COL_INFO::initialize_col_pack_info(TABLE_SHARE *table_share,
+                                                   uint32_t keynr) {
+  int32_t error = ENOSYS;
+  //
+  // set up the cp_info
+  //
+  assert_always(cp_info[keynr] == NULL);
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: nullptr

File: storage/tokudb/hatoku_cmp.cc [hunk 8]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+      if (error) {
+        goto exit;
+      }
+    }
+  }
+exit:
+  return error;
+}
+
+int32_t KEY_AND_COL_INFO::initialize_col_pack_info(TABLE_SHARE *table_share,
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: consider const ref for table_share

File: storage/tokudb/hatoku_cmp.cc [hunk 9]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+      bitmap_free(&key_filters[i]);
+    }
+    tokudb::memory::free(multi_ptr);
+  } else {
+    allocated = true;
+  }
+  return error;
+}
+
+int32_t KEY_AND_COL_INFO::initialize(TABLE_SHARE *table_share, TABLE *table,
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: Consider const ref for table_share and table

File: storage/tokudb/hatoku_cmp.cc [hunk 10]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+  //
+  // create the field lengths
+  //
+  multi_ptr = tokudb::memory::multi_malloc(
+      MYF(MY_WME + MY_ZEROFILL), &field_types,
+      (uint)(table_share->fields * sizeof(uint8_t)), &field_lengths,
+      (uint)(table_share->fields * sizeof(uint16_t)), &length_bytes,
+      (uint)(table_share->fields * sizeof(uint8_t)), &blob_fields,
+      (uint)(table_share->fields * sizeof(uint32_t)), NullS);
+  if (multi_ptr == NULL) {
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: nullptr

File: storage/tokudb/hatoku_cmp.cc [hunk 11]
```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
...
+      blob_fields(nullptr),
+      num_blobs(0),
+      num_offset_bytes(0) {
+  memset(&mcp_info, 0, sizeof(mcp_info));
+  memset(&cp_info, 0, sizeof(cp_info));
+}
+
+KEY_AND_COL_INFO::~KEY_AND_COL_INFO() { free(); }
+
+int32_t KEY_AND_COL_INFO::allocate(TABLE_SHARE *table_share) {
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.cc]: Consider const ref for table_share

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by george-lorch** [comment]
Made suggested changes, also wen a little deeper down the rabbit hole in refactoring and const ref changes.  New jenkins https://ps.cd.percona.com/view/8.0/job/percona-server-8.0-param/88/

File: storage/tokudb/hatoku_cmp.h [hunk 12]
```diff
@@ -100,9 +100,130 @@ typedef struct st_multi_col_pack_info {
...
+    return ret_val;
+  }
+
+ private:
+  bool allocated;
+
+ public:
   //
-  // bitmaps for each key. key_filters[i] is associated with the i'th
+  // bitmaps for each key..ey_filters[i] is associated with the i'th
```
Comment (laurynas-biveinis) [inline storage/tokudb/hatoku_cmp.h]: Oops :)

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: storage/tokudb/hatoku_cmp.h (hunk 12 above)
Comment (george-lorch) [reply storage/tokudb/hatoku_cmp.h]: what the heck did I do there :-\

Comment (laurynas-biveinis) [reply storage/tokudb/hatoku_cmp.h]: Probably tried to quit vi :)

Comment (george-lorch) [reply storage/tokudb/hatoku_cmp.h]: ZING!

**Comment by george-lorch** [comment]
Fixed typo, also fixed format issue pointed out by Travis clang-format-diff.

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by peterschwaller** [comment]
Dummy comment to trigger sync to Jira.  Please ignore.

**Pull request merged**